sources/scraping/helpers.py

```diff
@@ -0,0 +1,51 @@
+import logging
+import threading
+from typing import Any, Callable, Dict, List, Type, TypeVar
+
+from scrapy import Spider  # type: ignore
+from scrapy.crawler import CrawlerRunner  # type: ignore
+from twisted.internet import reactor
+
+from .types import BaseQueue
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+def init_scrapy_runner(  # type: ignore[no-untyped-def]
+    name: str,
+    start_urls: List[str],
+    spider: Type[Spider],
+    queue: BaseQueue[T],
+    settings: Dict[str, Any],
+    **kwargs,
+) -> None:
+    """Builds and prepares to run crawler with prepared configuration"""
+    runner = CrawlerRunner()
+    runner.crawl(
+        spider,
+        queue=queue,
+        name=f"{name}_spider",
+        start_urls=start_urls,
+        settings=settings,
+        **kwargs,
+    )
+
+    d = runner.join()
+    d.addBoth(lambda _: reactor.stop())  # type: ignore[attr-defined]
+    reactor.run()  # type: ignore[attr-defined]
```

Comment:
here you should
1. wait for scrapy to finish (all spiders stopped yielding)
2. make sure that all messages got consumed from the queue
3. close the queue at the very end